stm32: Add support for G0 series and add BOARD_G0B1RE board dfn.

AsensioL opened the pull request:
This commit adds support for the STM32G0 series of MCUs, and a board
definition for BOARD_G0B1RE .  This board has the REPL on UART2 which is
connected to the on-board ST-link USB-UART.
Closes #8292

AsensioL committed:
stm32: Add support for G0 MCUs.

This commit adds support for the STM32G0 series of MCUs.

Signed-off-by: Asensio Lorenzo Sempere <asensio.aerospace@gmail.com>

AsensioL committed:
stm32/boards/NUCLEO_G0B1RE: Add NUCLEO_G0B1RE board definition.

This commit adds a board definition for NUCLEO_G0B1RE. This board has the
REPL on UART2 which is connected to the on-board ST-link USB-UART.

Signed-off-by: Asensio Lorenzo Sempere <asensio.aerospace@gmail.com>

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -336,12 +336,18 @@ SRC_O += \
 	resethandler_m0.o \
 	shared/runtime/gchelper_m0.o
 else
+ifeq ($(MCU_SERIES),g0)
```
Instead of duplicating code, this could be rolled in to the `f0` block above (using `filter`).

dpgeorge commented on ports/stm32/boards/stm32g0xx_hal_conf_base.h:
```diff
@@ -0,0 +1,212 @@
+/*
```
please make this file look more like `ports/stm32/boards/stm32g4xx_hal_conf_base.h` in terms of style/formatting

dpgeorge commented on ports/stm32/dac.c:
```diff
@@ -63,6 +63,8 @@
 ///     dac = DAC(1)
 ///     dac.write_timed(buf, 400 * len(buf), mode=DAC.CIRCULAR)
+//------------------------------------------
+//#define MICROPY_HW_ENABLE_DAC 1 // Delete me
```
please remove this

dpgeorge commented on ports/stm32/mphalport.h:
```diff
@@ -21,6 +21,8 @@
 #elif defined(STM32WL)
 #define MICROPY_PLATFORM_VERSION "HAL1.1.0"
+#elif defined(STM32G0)
+#define MICROPY_PLATFORM_VERSION "HAL1.2.0"
```
please sort before G4

dpgeorge commented on ports/stm32/powerctrl.c:
```diff
@@ -286,7 +286,7 @@ int powerctrl_rcc_clock_config_pll(RCC_ClkInitTypeDef *rcc_init, uint32_t sysclk
 #endif
 
-#if !defined(STM32F0) && !defined(STM32L0) && !defined(STM32L4)
+#if !defined(STM32F0) && !defined(STM32L0) && !defined(STM32L4) && !defined(STM32G0)
```
please sort before L0

dpgeorge commented:
Thanks for the contribution, it looks good!

I do not yet have a G0 board.  What sort of testing did you do?  Did you test reading/writing the internal filesystem?

AsensioL commented:
As for the testing, I followed the [Writing Tests](https://docs.micropython.org/en/latest/develop/writingtests.html) instructions, and ran the tests on the board using `python run-tests.py --target minimal --device COM4 >> test_result.txt`.

The results, after filtering passed and skipped tests, are:
```
FAIL  extmod/uselect_poll_basic.py
FAIL  extmod/uselect_poll_udp.py
FAIL  extmod/usocket_udp_nonblock.py
FAIL  extmod/utime_res.py
FAIL  extmod/utime_time_ns.py
FAIL  extmod/vfs_fat_mtime.py
FAIL  micropython/import_mpy_native_gc.py
688 tests performed (19809 individual testcases)
681 tests passed
105 tests skipped: array_construct_endian btree1 btree_error {omitted}
7 tests failed: import_mpy_native_gc uselect_poll_basic uselect_poll_udp usocket_udp_nonblock utime_res utime_time_ns vfs_fat_mtime
```

AsensioL commented:
I just noticed that I had my editor set to use tabs instead of 4 spaces.
I'll fix that and push again later on a style-only commit.

AsensioL commented:
Regarding reading/writing the internal filesystem, because the STM32G0 is an ARM Cortex M0+, the NVIC register does not have a `NVIC->STIR` register (used by `storage_systick_callback(uint32_t ticks_ms)` in storage.c).

I did some research, and found [this article](https://www.sciencedirect.com/topics/engineering/interrupt-controller), where the first few paragraphs indicate that `NVIC->STIR = FLASH_IRQn;` can be replaced with `HAL_NVIC_SetPendingIRQ(FLASH_IRQn);`.

I tried it and I was able to successfully write/read to internal storage (see picture below).
![image](https://user-images.githubusercontent.com/36895998/156671459-0884ab49-aea4-4443-bf34-81488c7be3cb.png)

I haven't pushed these changes because, as the article explains, the `NVIC->STIR` can be readily modified from the User Space, but `HAL_NVIC_SetPendingIRQ(..)` requires additional changes.

AsensioL commented:
Oh, I forgot to mention, I'm using the STM32 G0 HAL as provided by ST (no modifications).
It is not as clear to me how/if I should branch and create a Pull Request for that too.
What would be the next step there?

dpgeorge commented:
> The results, after filtering passed and skipped tests, are:

OK, that is good, it's fine if only those 7 tests fail.

> I'll run tools/codeformat.py later and push again on a style-only commit.

There are still some code formatting issues to address.

> Regarding reading/writing the internal filesystem, because the STM32G0 is an ARM Cortex M0+, the NVIC register does not have a `NVIC->STIR` register (used by `storage_systick_callback(uint32_t ticks_ms)` in storage.c).

OK, so I suggest a patch like this:
```patch
--- a/ports/stm32/storage.c
+++ b/ports/stm32/storage.c
@@ -83,7 +83,11 @@ uint32_t storage_get_block_count(void) {
 static void storage_systick_callback(uint32_t ticks_ms) {
     if (STORAGE_IDLE_TICK(ticks_ms)) {
         // Trigger a FLASH IRQ to execute at a lower priority
+        #if __CORTEX_M == 0
+        NVIC_SetPendingIRQ(FLASH_IRQn);
+        #else
         NVIC->STIR = FLASH_IRQn;
+        #endif
     }
 }
```
That should hopefully get storage working on Cortex-M0 MCUs.

> Oh, I forgot to mention, I'm using the STM32 G0 HAL as provided by ST (no modifications), pulled directly through the tool in stm32lib.
> It is not as clear to me how/if I should branch and create a Pull Request for that too.
> What would be the next step there?

It's easiest for me to just add the G0 HAL to stm32lib, and update everything.  I can do that just before this PR gets merged.

dpgeorge commented on ports/stm32/Makefile:
```diff
@@ -324,13 +324,7 @@ SRC_O += \
-	shared/runtime/gchelper_m0.o
-else
-ifeq ($(MCU_SERIES),l0)
-CSUPEROPT = -Os # save some code space
```
This CSUPEROPT line needs to remain.

dpgeorge commented on ports/stm32/adc.c:
```diff
@@ -209,7 +209,7 @@ STATIC bool is_adcx_channel(int channel) {
     #elif defined(STM32F0) || defined(STM32F4) || defined(STM32F7)
     return IS_ADC_CHANNEL(channel);
-    #elif defined(STM32H7)
+    #elif defined(STM32H7) || defined(STM32G0)
```
please sort G0 before H7

dpgeorge commented on ports/stm32/adc.h:
```diff
@@ -42,7 +42,7 @@ static inline void adc_deselect_vbat(ADC_TypeDef *adc, uint32_t channel) {
         ADC_Common_TypeDef *adc_common;
 
-        #if defined(STM32F0) || defined(STM32WB)
+        #if defined(STM32F0) || defined(STM32WB) || defined(STM32G0)
```
please sort G0 before WB

dpgeorge commented on ports/stm32/dma.c:
```diff
@@ -113,7 +113,7 @@ static const DMA_InitTypeDef dma_init_struct_spi_i2c = {
     #if defined(STM32F4) || defined(STM32F7)
     .Channel = 0,
-    #elif defined(STM32H7) || defined(STM32L0) || defined(STM32L4)
+    #elif defined(STM32H7) || defined(STM32L0) || defined(STM32L4) || defined(STM32G0)
```
please sort G0 alphabetically before H7

dpgeorge commented on ports/stm32/dma.h:
```diff
@@ -28,7 +28,7 @@
 typedef struct _dma_descr_t dma_descr_t;
 
-#if defined(STM32F0) || defined(STM32F4) || defined(STM32F7) || defined(STM32H7)
+#if defined(STM32F0) || defined(STM32F4) || defined(STM32F7) || defined(STM32H7) || defined(STM32G0)
```
please sort G0 before H7

dpgeorge commented on ports/stm32/machine_adc.c:
```diff
@@ -267,7 +267,7 @@ STATIC void adc_config_channel(ADC_TypeDef *adc, uint32_t channel, uint32_t samp
     #endif
 
-    #if defined(STM32F0) || defined(STM32L0)
+    #if defined(STM32F0) || defined(STM32L0) || defined(STM32G0)
```
please sort G0

dpgeorge commented on ports/stm32/mpconfigboard_common.h:
```diff
@@ -319,6 +319,16 @@
+#define PYB_EXTI_NUM_VECTORS (22) // previously 23
+#define MICROPY_HW_MAX_I2C (3)
+#define MICROPY_HW_MAX_TIMER (17)
+#define MICROPY_HW_MAX_UART (6) // UART1-6 + LPUART1
```
is this comment correct, since there are 2 LPUART's?  maybe just remove the comment?

dpgeorge commented on ports/stm32/powerctrlboot.c:
```diff
@@ -104,6 +104,77 @@ void SystemClock_Config(void) {
+    // Set flash latency to 2 because SYSCLK > 48MHz
+    FLASH->ACR = (FLASH->ACR & ~0x7) | 0x2;
+
+    #if MICROPY_HW_CLK_USE_HSI
```
Is there a reason you didn't implement HSE support?  Just for simplicity?

AsensioL replied:
Yes, I just wanted to go for a minimal implementation.
I figured I could add HSE later.

dpgeorge replied:
OK

dpgeorge commented on ports/stm32/rtc.c:
```diff
@@ -742,6 +746,8 @@ mp_obj_t pyb_rtc_wakeup(size_t n_args, const mp_obj_t *args) {
         #elif defined(STM32H7)
         EXTI_D1->PR1 = 1 << EXTI_RTC_WAKEUP;
+        #elif defined(STM32G0)
+        // TODO: No need to clear this, Confirm
```
does this need to be checked?

dpgeorge commented on ports/stm32/uart.c:
```diff
@@ -497,6 +522,32 @@ bool uart_init(pyb_uart_obj_t *uart_obj,
+            #if defined(STM32G0)
+            irqn = USART2_LPUART2_IRQn;
+            #else
+            irqn = USART2_IRQn;
```
I guess this line hasn't been tested on hardware, so best to remove it, just support the G0 for LPUART2 for now.

chrismas9 commented:
I can confirm that the change to `storage.c` works for f0. Tested on NUCLEO_F091RC. Tested creating LFS1 filesystem, writing and reading main.py. I have a PR in the works.

AsensioL commented:
> I can confirm that the change to `storage.c` works for f0. Tested on NUCLEO_F091RC. Tested creating LFS1 filesystem, writing and reading main.py. I have a PR in the works.

Thanks for the feedback! I'm glad it is working on F0 too.

AsensioL commented:
> > The results, after filtering passed and skipped tests, are:
> 
> OK, that is good, it's fine if only those 7 tests fail.
> 
> > I'll run tools/codeformat.py later and push again on a style-only commit.
> 
> There are still some code formatting issues to address.
> 
> > Regarding reading/writing the internal filesystem, because the STM32G0 is an ARM Cortex M0+, the NVIC register does not have a `NVIC->STIR` register (used by `storage_systick_callback(uint32_t ticks_ms)` in storage.c).
> 
> OK, so I suggest a patch like this:
> 
> ```diff
> --- a/ports/stm32/storage.c
> +++ b/ports/stm32/storage.c
> @@ -83,7 +83,11 @@ uint32_t storage_get_block_count(void) {
>  static void storage_systick_callback(uint32_t ticks_ms) {
>      if (STORAGE_IDLE_TICK(ticks_ms)) {
>          // Trigger a FLASH IRQ to execute at a lower priority
> +        #if __CORTEX_M == 0
> +        NVIC_SetPendingIRQ(FLASH_IRQn);
> +        #else
>          NVIC->STIR = FLASH_IRQn;
> +        #endif
>      }
>  }
> ```
> 
> That should hopefully get storage working on Cortex-M0 MCUs.
> 
> > Oh, I forgot to mention, I'm using the STM32 G0 HAL as provided by ST (no modifications), pulled directly through the tool in stm32lib.
> > It is not as clear to me how/if I should branch and create a Pull Request for that too.
> > What would be the next step there?
> 
> It's easiest for me to just add the G0 HAL to stm32lib, and update everything. I can do that just before this PR gets merged.

Is there anything else I should do?
I know the Pull Request is a bit messy at this point, so I assume I should create a cleaner one once all the changes on this one are approved.
At this point I would assume I would have 3 commits:
* Changes to storage.c to add support for Cortex-M0 MCUs
* Add support for the STM32G0 series of MCUs,
* Add board definition for NUCLEO_G0B1RE 
definition for BOARD_G0B1RE 

 I will get the commit message formatting right this time.

dpgeorge commented:
> Is there anything else I should do? I know the Pull Request is a bit messy at this point, so I assume I should create a cleaner one once all the changes on this one are approved. At this point I would assume I would have 3 commits:
> 
>     * Changes to storage.c to add support for Cortex-M0 MCUs
> 
>     * Add support for the STM32G0 series of MCUs,
> 
>     * Add board definition for NUCLEO_G0B1RE
>       definition for BOARD_G0B1RE

Yes, please, if you can reorganise/squash the commits here into those 3 commits that would be great!  That makes it easier for me to merge.

You can keep this PR and just force-push the changes.

AsensioL committed:
stm32/storage: Add support for internal storage on Cortex-M0 MCUs.

This implements self-triggering of the Flash NVIC interrupt on Cortex-M0
devices, which allows enabling internal storage on those MCUs.

Signed-off-by: Asensio Lorenzo Sempere <asensio.aerospace@gmail.com>

AsensioL commented:
I think I got it right this time. Let me know if anything needs to be changed.

Thank you for all the reviews!

napalm51 commented:
Hello AsenioL and thanks for supporting STM32G0.
I've tried to compile with:
```
make V=1 BOARD=NUCLEO_G0B1RE
make: ***  No rule to make target "build-NUCLEO_G0B1RE/modstm_qstr.h", needed by "build-NUCLEO_G0B1RE/genhdr/qstrdefs.generated.h".  Stop.
 
```
Looking on the forum I see that the problem is maybe related to "hal and cmsis files has not update", but I've no mean to fix that. Can you kindly help with that? Thank you

AsensioL commented:
It is likely that the problem you're running into is that you're missing those files (CMSIS and HAL) since they have not yet been included the **stm32lib** submodule.

In that case, what you need to do is:
1. Clone the CMSIS + HAL from the STM32G0 github ([link](https://github.com/STMicroelectronics/STM32CubeG0)) to _lib/stm32lib_ inside your local copy of the project. This should create a new folder with a name like `STM32CubeG0`
2. Go to _lib/stm32lib_ on your terminal and run `fetch_from_upstream.sh STM32G0 STM32CubeG0`.

The script should create all the files you are missing. Try recompiling and see if it works.

AsensioL commented:
Code has been rebased and is ready to be reviewed.
Commit messages have also been fixed.
There are no conflicts now!

dpgeorge commented:
Thanks very much for the hard work on this PR!  The commits are nice and clean.  I tested building PYBV10, PYBD_SF6 and NUCLEO_WB55 and they produce identical firmware before and after this PR, which is very good, it means that the commits here do not touch other MCUs.

I updated stm32lib to include G0 at v1.5.1.  This builds against the NUCLEO_G0B1RE board but I did not test it.

Merged in 3587d8e80853051b84da25e8d9abc3422bc6f57f through 9d08eb02474fcace0a81dd3518b938afbf1486eb
